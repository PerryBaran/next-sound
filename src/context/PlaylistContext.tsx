'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
