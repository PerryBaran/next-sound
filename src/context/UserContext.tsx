'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
