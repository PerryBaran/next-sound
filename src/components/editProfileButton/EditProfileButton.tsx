"use client"

import { useUserContext } from "@/context/UserContext"
import Link from "next/link"

interface Props {
  userId: string
}

export default function EditProfileButton({ userId }: Props) {
  const {
    user: { id }
  } = useUserContext()

  if (id !== userId) return null

  return <Link href={`/edit/profile`}>Edit Profile</Link>
}
