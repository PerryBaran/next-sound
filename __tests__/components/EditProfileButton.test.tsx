import { render, screen } from "@testing-library/react"
import { createContext, useContext } from "react"
import EditProfileButton from "../../src/components/editProfileButton/EditProfileButton"
import * as UserContext from "../../src/context/UserContext"

describe("EditProfileButton", () => {
  const props = {
    userId: "user-id"
  }

  beforeEach(() => {
    const mockContext = createContext({
      user: {
        id: props.userId,
        name: "test-name"
      },
      handleLogin: () => {}
    })

    jest
      .spyOn(UserContext, "useUserContext")
      .mockImplementation(() => useContext(mockContext))
  })

  test("if id in context doesn't equal userId, return null", () => {
    const { container } = render(<EditProfileButton userId="different" />)

    expect(container).toBeEmptyDOMElement()
  })

  test("if id in context is equal to userId, renders a link", () => {
    render(<EditProfileButton {...props} />)

    expect(screen.getByText(/edit profile/i)).toHaveAttribute(
      "href",
      "/edit/profile"
    )
  })

  describe("snapshots", () => {
    test("id in context doesn't equal userId", () => {
      const { asFragment } = render(<EditProfileButton userId="different" />)

      expect(asFragment()).toMatchSnapshot()
    })

    test("id in context equals userId", () => {
      const { asFragment } = render(<EditProfileButton {...props} />)

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
