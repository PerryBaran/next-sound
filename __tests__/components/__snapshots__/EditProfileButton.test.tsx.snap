// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfileButton snapshots id in context doesn't equal userId 1`] = `<DocumentFragment />`;

exports[`EditProfileButton snapshots id in context equals userId 1`] = `
<DocumentFragment>
  <a
    href="/edit/profile"
  >
    Edit Profile
  </a>
</DocumentFragment>
`;
